### PR TITLE
Feature: !techcrunch command which posts a message with 5 top-most TechCrunch articles in the Latest News section

### PR DESCRIPTION
**Pull request by gaenrique**
This command webscrapes techcrunch.com and posts a message containing the top 5 articles in the Latest News section. The message which the bot sends also makes the headlines clickable (linking to the article).

I believe that if people enjoy this command, it would be a nice bot command to run automatically once a day.

### CODE REVIEW

**Review by nicklambourne — COMMENTED** [review]
Have you considered using the RSS feed, rather than scraping the HTML? TC doesn't change much, but the RSS feed is sure to be less brittle than the site.

**Comment by gaenrique** [comment]
@nicklambourne I have not had a look at that. I'll make sure to give it a go when I have time. Thanks for the feedback

**Comment by gaenrique** [comment]
Jenkins seems to be failing since build 795. I don't know what the problem is.

**Comment by nicklambourne** [comment]
> uqcsbot/scripts/news.py:6:1: F401 'requests' imported but unused
> uqcsbot/scripts/news.py:7:1: F401 'bs4.BeautifulSoup' imported but unused
> uqcsbot/scripts/news.py:14:42: W291 trailing whitespace
> uqcsbot/scripts/news.py:18:31: E203 whitespace before ':'
> uqcsbot/scripts/news.py:18:61: E203 whitespace before ':'
> uqcsbot/scripts/news.py:24:55: W291 trailing whitespace
> uqcsbot/scripts/news.py:39:1: W293 blank line contains whitespace

**Comment by gaenrique** [comment]
Thanks! I didn't how to find the errors but I found them now.

**Comment by gaenrique** [comment]
Alright now for some reason the trivia tests fail even though I didn't touch them. I'll try and find out what is wrong sorry

**Comment by nicklambourne** [comment]
There's a good chance that's not your fault, some of our tests are flakey (and by ours I mostly mean Bradley).

**Comment by nicklambourne** [comment]
Also, no need to be sorry. I for one am thrilled someone's working on the bot again. The CI checks exist so we know what's wrong, they exist independently of the bot actually running in Slack, so there's no issue if they fail during development.

**Comment by gaenrique** [comment]
I appreciate the support. First time contributing to open source so I'm new to all this

**Comment by gaenrique** [comment]
One thing is for sure though. I do not understand this trailing whitespace error

**Comment by gaenrique** [comment]
Weird random errors keep occurring so I might just give it a break and try again later

**Comment by nicklambourne** [comment]
@bradleysigma unless you can fix those tests I'm going to remove !yelling.

**Comment by nicklambourne** [comment]
@gaenrique Give us a day or so to sort things out with Jenkins, we'll get it merged, I'm just at work at the moment.

**Comment by nicklambourne** [comment]
@kentonlam are you still technical czar?

**Comment by gaenrique** [comment]
@nicklambourne sounds good. I'll just give it one more shot as the last error was just an over-indentation on my file

File: uqcsbot/scripts/news.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    return (news[index]['title'], news[index]['link'])
+
+@bot.on_command("news")
+@loading_status
+def handle_news(command: Command) -> None:
+    """
+    Prints the 5 top-most articles in the Latest News Section of TechCrunch
+    using RSS feed
+    """
+    message = "------------------------- Latest News from TechCrunch ---------------------------\n"
```
Comment (kentonlam) [inline uqcsbot/scripts/news.py:31]: Formatting nit, can the heading be bold like `**Example Title**` instead? A link to techcrunch.com would also be nice.

File: uqcsbot/scripts/news.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    news = get_tech_crunch_data()
+    for i in range(ARTICLES_TO_POST):
+        title, url = get_data_from_article(news, i)
+        # Formats message a clickable headline which links to the article
+        message += f"<{url}|{title}>\n\n"
```
Comment (kentonlam) [inline uqcsbot/scripts/news.py:36]: Formatting: Can these be bullet points? Prefix each line with `-` and a space.

File: uqcsbot/scripts/news.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+def get_data_from_article(news: List[Dict[str, str]], index: int) -> Tuple[str, str]:
+    """
+    Returns the title of the article and the link
+
+    Tuple returned: (title, url)
+    """
+    return (news[index]['title'], news[index]['link'])
+
+@bot.on_command("news")
```
Comment (kentonlam) [inline uqcsbot/scripts/news.py:24]: This feels a bit too specific for a `!new` command. Maybe in future if it supported multiple news sources, but I think `!techcrunch` would be more suitable at the moment.

File: uqcsbot/scripts/news.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+from uqcsbot import bot, Command
+from uqcsbot.utils.command_utils import loading_status
+from typing import Tuple, Dict, List
+
+import feedparser
```
Comment (kentonlam) [inline uqcsbot/scripts/news.py:5]: Does this need to be added as a dependency somewhere? (I think yes, but the tests seem to be passing 🤔)

**Review by kentonlam — CHANGES_REQUESTED** [review]
Thanks for the changes! 😃

I thought we found and fixed the !yelling problem with some checks. @bradleysigma? 👀

File: uqcsbot/scripts/news.py (hunk 1 above)
Comment (gaenrique) [reply uqcsbot/scripts/news.py:31]: How do you enable Markdown formatting so that the heading is bold. Simply using ** does not seem to work for me

Comment (kentonlam) [reply uqcsbot/scripts/news.py:31]: My bad, Slack uses single asterisks for bold: https://api.slack.com/reference/surfaces/formatting

File: uqcsbot/scripts/news.py (hunk 3 above)
Comment (gaenrique) [reply uqcsbot/scripts/news.py:24]: I also think this is a good idea. I'll link it to the `!techcrunch` command for now. But in the future when I have more time I would like to accept arguments such as `!news techcrunch` and `!news 9news' etc and post headlines for chosen news sites

File: uqcsbot/scripts/news.py (hunk 4 above)
Comment (nicklambourne) [reply uqcsbot/scripts/news.py:5]: `feedparser` is already listed explicitly in `setup.py`, so no worries there.

**Comment by bradleysigma** [comment]
> 
> 
> @bradleysigma unless you can fix those tests I'm going to remove !yelling.

Remove the tests or remove the feature?

**Comment by gaenrique** [comment]
Changes have been implemented. Feel free to give any additional feedback if you would like anything else changed :)

**Comment by nicklambourne** [comment]
@bradleysigma the feature. You may remove the offending tests as a half-measure if you so choose.

**Review by nicklambourne — APPROVED** [review]
LGTM, let me know what you think @kentonlam

**Comment by gaenrique** [comment]
@nicklambourne I actually just realised that I forgot to add an error message if the RSS feed can't be accessed so I'll just finish that up real quick

File: uqcsbot/scripts/techcrunch.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+def handle_news(command: Command) -> None:
+    """
+    Prints the 5 top-most articles in the Latest News Section of TechCrunch
+    using RSS feed
+    """
+    message = f"*------------------- Latest News from <{TECHCRUNCH_URL}|_TechCrunch_> " \
+              f":newspaper: ---------------------*\n"
+    news = get_tech_crunch_data()
+    if news is None:
+        bot.post_message(command.channel_id, "There was an error accessing "
```
Comment (kentonlam) [inline uqcsbot/scripts/techcrunch.py:39]: Return after posting error message?

File: test/test_techcrunch.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import feedparser
+from typing import List, Dict
+
+def get_local_techcrunch_data() -> List[Dict[str, str]]:
+    """
+    Returns the RSS data from local mock copy of techcrunch xml.
+    """
+    try:
+        rawdata = open("test/techcrunch.xml").read()
```
Comment (kentonlam) [inline test/test_techcrunch.py:12]: Can we use `with open(...)` instead? This makes sure to clean up the opened file.

File: test/test_techcrunch.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        data = feedparser.parse(rawdata).entries
+    except Exception as e:
+        raise e
+        return (0, "")
```
Comment (kentonlam) [inline test/test_techcrunch.py:16]: This is superfluous. return will never be reached and there's no need to catch then immediately re-raise an exception.

File: uqcsbot/scripts/techcrunch.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+TECHCRUNCH_URL = "https://techcrunch.com"
+
+def get_tech_crunch_data() -> List[Dict[str, str]]:
+    """
+    Returns data from TechCrunch RSS feed
+    """
+    data = feedparser.parse(RSS_URL)
+    if data.status != 200:
+        return None
+    return feedparser.parse(RSS_URL).entries
```
Comment (kentonlam) [inline uqcsbot/scripts/techcrunch.py:18]: Nit: can use the data variable here.

**Review by kentonlam — CHANGES_REQUESTED** [review]
Just a few small changes and style things. 🙂

File: test/test_techcrunch.py (hunk 7 above)
Comment (gaenrique) [reply test/test_techcrunch.py:16]: So would you like me to print something before raising the exception?

Comment (kentonlam) [reply test/test_techcrunch.py]: I'd just let the exception throw, but printing something is fine too.

Comment (nicklambourne) [reply test/test_techcrunch.py]: I think Kenton is just suggesting that you either a) don't catch it at all, or b) catch it, send a message indicating the error to Slack and then exit without re-raising. My preference would be for b).

**Review by kentonlam — APPROVED** [review]
Thanks for fixing the style! You can merge it whenever you're ready.

File: test/test_techcrunch.py (hunk 7 above)
Comment (gaenrique) [reply test/test_techcrunch.py]: @nicklambourne this try block is in the test file though, so there's no point sending a message to Slack right? Or is there a specific channel to send test fail messages to?

Comment (kentonlam) [reply test/test_techcrunch.py]: Generally, sending an error message is good but this is for testing so we can't do that here.

**Comment by gaenrique** [comment]
> Thanks for fixing the style! You can merge it whenever you're ready.

Final questions since I'm new to all this. So now that it is all approved, do I merge it or do you guys have to merge?

**Comment by kentonlam** [comment]
Fairly certain we have to merge because of permissions.

**Pull request merged**